### ISSUE TITLE
Dapper v1.40 Transacton bug

### ISSUE THREAD

**Issue by tomexou** [StackExchange/dapper-dot-net#276]
Considers the snippet below, the transaction's connect reference will be null when an exception was thrown.

```
var trans = con.BeginTransaction();

try {

  con.Execute(sql, params, trans);  // Let this throw an exception.

  trans.Commit();
}
catch
{
  //Exception => "Conversion failed when converting the nvarchar value '10.5' to data type int."

  if (trans.Connection == null)
  {
     // WHY?
     // At here, the trans.Connection will be null when an exception thrown.
     // It will make the following Rollback() without any efforts.
  }

  trans.Rollback();
}
```

**Comment by mgravell** [comment]
why would trans.Connection be null?

**Comment by tomexou** [comment]
I gave the dbParameter with "10.5" string into an int column, then it throws an exception.
Error msg is: "Conversion failed when converting the nvarchar value '10.5' to data type int."

I have no idea why the transaction will be null after mismatch type exception.

**Comment by mgravell** [comment]
What rdbms / ado.net provider is this? I.e. what is dapper talking to?

**Comment by tomexou** [comment]
SQL Server 2014 LocalDB, and the Provider is System.Data.SqlClient in .NET 4.0 runtime.

**Comment by tomexou** [comment]
c# - SqlTransaction after catch transaction connection is null - Stack Overflow
http://stackoverflow.com/questions/9839631/sqltransaction-after-catch-transaction-connection-is-null

>>
Ran into a similar issue. In my case it was happening for a specific SqlException. Most exceptions would be caught and handled just fine, but whenever I got a conversion error (such as trying to convert a string to a number) it would automatically end the transaction.

It is so odd. There is no good solution until now.
